src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/NuGetDependenciesSubTreeProvider.cs

```diff
@@ -216,96 +227,259 @@ public override IDependencyNode GetDependencyNode(DependencyNodeId id)
         }
 
         protected override DependenciesChange ProcessDependenciesChanges(
-                                                    IProjectSubscriptionUpdate projectSubscriptionUpdate,
-                                                    IProjectCatalogSnapshot catalogs)
+                                                     IProjectSubscriptionUpdate projectSubscriptionUpdate,
+                                                     IProjectCatalogSnapshot catalogs)
         {
-            var changes = projectSubscriptionUpdate.ProjectChanges;
-            var dependenciesChange = new DependenciesChange();
+            var dependenciesChange = new DependenciesChange();            
+            // take a snapshot for current top level tree nodes
+            var rootNodes = new HashSet<IDependencyNode>(RootNode.Children);
 
             lock (_snapshotLock)
             {
-                var newDependencies = new HashSet<DependencyMetadata>();
-                foreach (var change in changes.Values)
+                var unresolvedChanges = ProcessUnresolvedChanges(projectSubscriptionUpdate);
+                var resolvedChanges = ProcessResolvedChanges(projectSubscriptionUpdate, rootNodes);
+
+                // merge changes
+                // remove
+                foreach (var metadata in unresolvedChanges.RemovedNodes)
                 {
-                    if (!change.Difference.AnyChanges)
+                    var itemNode = rootNodes.FirstOrDefault(
+                                    x => x.Id.ItemSpec.Equals(metadata.ItemSpec, StringComparison.OrdinalIgnoreCase));
+                    if (itemNode != null)
                     {
-                        continue;
+                        dependenciesChange.RemovedNodes.Add(itemNode);
                     }
+                }
 
-                    foreach (string removedItemSpec in change.Difference.RemovedItems)
+                foreach (var metadata in resolvedChanges.RemovedNodes)
+                {
+                    var itemNode = rootNodes.FirstOrDefault(
+                                    x => x.Id.ItemSpec.Equals(metadata.ItemSpec, StringComparison.OrdinalIgnoreCase));
+                    if (itemNode != null)
                     {
-                        CurrentSnapshot.RemoveDependency(removedItemSpec);
+                        dependenciesChange.RemovedNodes.Add(itemNode);
+                    }
+                }
 
-                        var itemNode = RootNode.Children.FirstOrDefault(
-                                        x => x.Id.ItemSpec.Equals(removedItemSpec, StringComparison.OrdinalIgnoreCase));
-                        if (itemNode != null)
-                        {
-                            dependenciesChange.RemovedNodes.Add(itemNode);
-                        }
+                // update
```

Comment:
line 278 unresolvedChanges.UpdatedNodes.Remove(unresolvedMatch);, remove unresolved element if resolved was added.

---
In reply to: [85853361](https://github.com/dotnet/roslyn-project-system/pull/675#discussion_r85853361) [](ancestors = 85853361)